include/glow/Importer/ModelLoader.h

```diff
@@ -0,0 +1,125 @@
+/**
+ * Copyright (c) 2017-present, Facebook, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef GLOW_IMPORTER_MODELLOADER_H
+#define GLOW_IMPORTER_MODELLOADER_H
+
+#include "glow/Base/Tensor.h"
+#include "glow/Graph/Graph.h"
+
+#include "llvm/ADT/ArrayRef.h"
+#include "llvm/Support/raw_ostream.h"
+
+#include <google/protobuf/text_format.h>
+
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+namespace glow {
+
+extern unsigned NCHW2NHWC[4];
+extern unsigned NHWC2NCHW[4];
+
+/// Returns true iff all elements of \p a are the same.
+bool arrayIsConstant(const llvm::ArrayRef<size_t> a);
+
+/// Prints a single serialized protocol buffer node. This method is useful for
+/// debugging the network and printing errors.
+template <typename T>
+void unexpectedNodeError(const T &node, llvm::StringRef message) {
+  std::string str;
+  google::protobuf::TextFormat::PrintToString(node, &str);
+  llvm::outs() << message << "\n" << str << "\n";
+}
+
+/// Reads a single integer.
+template <typename T> static int loadInt(const T *arg) {
+  assert(arg->has_i() && "Node has no Int value");
+  return arg->i();
+}
+
+/// Reads a single float.
+template <typename T> static float loadFloat(const T *arg) {
+  assert(arg->has_f() && "Node has no float value");
+  return arg->f();
+}
+
+/// Reads a single string.
+template <typename T> static const std::string &loadStr(const T *arg) {
+  assert(arg->has_s() && "Node has no str value");
+  return arg->s();
+}
+
+/// Load the 'shape' record into a vector of sizes.
+template <typename T> std::vector<size_t> getShape(const T *arg) {
+  std::vector<size_t> dim;
+  for (auto i : arg->ints()) {
+    dim.push_back(i);
+  }
+  return dim;
+}
+
+/// Loads array and checks that all elements are the same. Returns array[0].
+template <typename T> size_t getConstantArrayHead(const T *arg) {
+  auto dim = getShape(arg);
+  assert(arrayIsConstant(dim) &&
+         "Only equal values along each dimensions are supported");
+  return dim[0];
+}
+
+/// Loads model: graph and weights.
+class ModelLoader {
+protected:
+  /// The graph that we are constructing.
+  Function &G_;
+  /// Saves network nodes by name.
+  std::unordered_map<std::string, Node *> nodeByName_;
+  /// A list of weight tensors indexed by name.
+  std::unordered_map<std::string, Tensor *> tensors_;
```

Comment:
Is there a reason to not make this:

  std::unordered_map<std::string, std::unique_ptr<Tensor>> tensors_;

Reply:
@compnerd Because the map does not own the tensors.

Reply:
@nadavrot hmm, so the Loader deletes the tensors explicitly, so the lifetime is associated with the loader.  The tensors are managed by the map.  So, it does make some sense to have the map own the pointer for the tensor.

Who exactly owns the tensors?

Reply:
It seems to me that Loader owns Tensors, but not Nodes. I'd switch to `unique_ptr` in follow up PR. Maybe I'm missing something.